hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/GrpcBlockItemWriter.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.blocks.impl;
+
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.hedera.hapi.block.protoc.BlockStreamServiceGrpc;
+import com.hedera.hapi.block.protoc.PublishStreamRequest;
+import com.hedera.hapi.block.protoc.PublishStreamResponse;
+import com.hedera.hapi.block.stream.protoc.BlockItem;
+import com.hedera.node.app.blocks.BlockItemWriter;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Implements the bidirectional streaming RPC for the publishBlockStream rpc in BlockStreamService
+ * See <a href="https://grpc.io/docs/languages/java/basics/">gRPC Basics</a>
+ */
+public class GrpcBlockItemWriter implements BlockItemWriter {
+
+    private static final Logger logger = LogManager.getLogger(GrpcBlockItemWriter.class);
+
+    private BlockStreamServiceGrpc.BlockStreamServiceStub asyncStub;
+    private StreamObserver<PublishStreamRequest> requestObserver;
+
+    /** The state of this writer */
+    private GrpcBlockItemWriter.State state;
+
+    private enum State {
+        UNINITIALIZED,
+        OPEN,
+        CLOSED
+    }
+
+    @Override
+    public void openBlock(long blockNumber) {
+        if (state != GrpcBlockItemWriter.State.UNINITIALIZED)
+            throw new IllegalStateException("Cannot initialize a GrpcBlockItemWriter twice");
+
+        if (blockNumber < 0) throw new IllegalArgumentException("Block number must be non-negative");
+
+        asyncStub = createBlockNodeAsyncStub();
+
+        StreamObserver<PublishStreamRequest> requestObserver =
+                asyncStub.publishBlockStream(new StreamObserver<PublishStreamResponse>() {
+                    @Override
+                    public void onNext(PublishStreamResponse value) {
+                        // TODO (driley) Handle EndofStream responses, and if Single Item Ack is enabled via config then
+                        // implement that as well.
+                        // However that would require keeping track of the hashes of the items sent and then verifying
+                        // before
```

Comment:
Note, itemAck is a debug option, and consensus node is free to accept-then-ignore single item acknowledgement.
End of Stream and Block acknowledgement _must_ be implemented, however.
The most complex option is probably "BEHIND".
Protocol is not fully set (suggestions welcome), but an initial description is in [this story](https://github.com/hashgraph/hedera-block-node/issues/120) to design the final protocol.

Reply:
Thanks for the additional information! Will take a look at the initial description.